src/os/bluestore/Writer.h

```diff
@@ -0,0 +1,214 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2023 IBM
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ */
+
+#ifndef BLUESTORE_WRITER
+#define BLUESTORE_WRITER
+
+#include "BlueStore.h"
+#include "Allocator.h"
+
+class BlueStore::Writer {
+public:
+  using exmp_it = extent_map_t::iterator;
+  using P = BlueStore::printer;
+
+  // Data that is to be put to object.
+  struct blob_data_t {
+    //uint32_t location;        // There is no need for each chunk to have separate location.
+    uint32_t real_length;       // Size of object data covered by this chunk. Same as object_data.length().
+    uint32_t compressed_length; // Size of compressed representation. 0 or disk_data.length().
+    bufferlist disk_data;       // Bitstream to got o disk. Its either same as object_data,
+                                // or contains compressed data. Block aligned.
+    bufferlist object_data;     // Object data. Needed to put into caches.
+    bool is_compressed() const {return compressed_length != 0;}
+  };
+  using blob_vec = std::vector<blob_data_t>;
+  struct blob_data_printer {
+    const blob_vec& blobs;
+    uint32_t base_position;
+    blob_data_printer(const blob_vec& blobs, uint32_t base_position)
+    : blobs(blobs), base_position(base_position) {}
+  };
+
+  struct write_divertor {
+    virtual ~write_divertor() = default;
+    virtual void write(
+      uint64_t disk_offset, const bufferlist& data, bool deferred) = 0;
+  };
+  struct read_divertor {
+    virtual ~read_divertor() = default;
+    virtual bufferlist read(uint32_t object_offset, uint32_t object_length) = 0;
+  };
+  Writer(BlueStore* bstore, TransContext* txc, WriteContext* wctx, OnodeRef o)
+    :bstore(bstore), txc(txc), wctx(wctx), onode(o) {
+      pp_mode = debug_level_to_pp_mode(bstore->cct);
+    }
+public:
+  void do_write(
+    uint32_t location,
+    bufferlist& data
+  );
+
+  void debug_iterate_buffers(
+    std::function<void(uint32_t offset, const bufferlist& data)> data_callback
+  );
+
+  write_divertor* test_write_divertor = nullptr;
+  read_divertor* test_read_divertor = nullptr;
+  std::vector<BlobRef> pruned_blobs;
+  std::set<SharedBlobRef> shared_changed;
```

Comment:
We can apparently use txc::shared_blobs directly and save an expensive insertion into the intermediate set.

Reply:
Done.